subsys/bluetooth/host/conn.c

```diff
@@ -1311,15 +1312,40 @@ struct bt_conn *bt_conn_ref(struct bt_conn *conn)
 void bt_conn_unref(struct bt_conn *conn)
 {
 	atomic_val_t old;
+	bool deallocated;
+	enum bt_conn_type conn_type;
+	uint8_t conn_role;
+	uint16_t conn_handle;
+
+	__ASSERT(conn, "Invalid connection reference");
+
+	/* Storing parameters of interest so we don't access the object
+	 * after decrementing its ref-count
+	 */
+	conn_type = conn->type;
+	conn_role = conn->role;
+	conn_handle = conn->handle;
 
 	old = atomic_dec(&conn->ref);
+	/* Prevent from accessing connection object */
+	conn = NULL;
+	deallocated = (atomic_get(&old) == 1);
 
-	LOG_DBG("handle %u ref %ld -> %ld", conn->handle, old, atomic_get(&conn->ref));
+	LOG_DBG("handle %u ref %ld -> %ld", conn_handle, old, (old - 1));
 
 	__ASSERT(old > 0, "Conn reference counter is 0");
 
-	if (IS_ENABLED(CONFIG_BT_PERIPHERAL) && conn->type == BT_CONN_TYPE_LE &&
-	    conn->role == BT_CONN_ROLE_PERIPHERAL && atomic_get(&conn->ref) == 0) {
+#if defined(CONFIG_BT_CONN)
```

Comment:
is this ifdef necessary?

Reply:
1. bt_conn_unref() is implemented even when CONFIG_BT_CONN is not set.
2. The helper function `notify_recycled_conn_slot()` is implemented within an #ifdefined CONFIG_BT_CONN, as the other helper notifiers do.
3. This combination causes a build-time failure on applications where CONFIG_BT_CONN is not enabled.

I do agree, however, that the code may throw-off the reader, as it isn't clear the additional dependency guard.

Question: Why do we have some areas in conn.c guarded with `CONFIG_BT_CONN` and some that don't? I would have thought the whole conn.c file either is required or it isn't. That's where my confusion comes from.

Suggestion: I can refactor the code to remove point No. 2, to define all the required functionality to notify `recycled()`. Note: this would mostly be for semantical purposes, as `bt_conn_cb_register()` is not defined if `CONFIG_BT_CONN` is not enabled.

Reply:
ah I see, this is the old "C sux", been there. I'll resolve the comment then.
Note that you can "solve" - if you can call it that - the problem by defining it that way:
```c
static void notify_recycled_conn_slot(void)
{
#ifdef CONFIG_BT_CONN
	for (struct bt_conn_cb *cb = callback_list; cb; cb = cb->_next) {
		if (cb->recycled) {
			cb->recycled();
		}
	}

	STRUCT_SECTION_FOREACH(bt_conn_cb, cb) {
		if (cb->recycled) {
			cb->recycled();
		}
	}
#else
#endif
}
```

Reply:
I like that suggestion. Not sure if we should leave the iterable section out or not, as technically one could technically define a `BT_CONN_CB_DEFINE()` listener irrespective of `bt_conn_cb_register()`.

In any case, I'll apply this change to improve readibility on `bt_conn_unref()`

Reply:
Addressed.